beacon/goclient/proposer.go

```diff
@@ -123,6 +174,116 @@ func (gc *goClient) SubmitBeaconBlock(block *spec.VersionedBeaconBlock, sig phas
 	return gc.client.SubmitBeaconBlock(gc.ctx, signedBlock)
 }
 
+func (gc *goClient) SubmitValidatorRegistration(pubkey []byte, feeRecipient bellatrix.ExecutionAddress, sig phase0.BLSSignature) error {
+	pk := phase0.BLSPubKey{}
+	copy(pk[:], pubkey)
+	signedReg := &api.VersionedSignedValidatorRegistration{
+		Version: spec.BuilderVersionV1,
+		V1: &eth2apiv1.SignedValidatorRegistration{
+			Message: &eth2apiv1.ValidatorRegistration{
+				FeeRecipient: feeRecipient,
+				// TODO: This is a reasonable default, but we should probably make this configurable.
+				//       Discussion here: https://github.com/ethereum/builder-specs/issues/17
+				GasLimit:  30_000_000,
+				Timestamp: gc.network.GetSlotStartTime(gc.network.GetEpochFirstSlot(gc.network.EstimatedCurrentEpoch())),
+				Pubkey:    pk,
+			},
+			Signature: sig,
+		},
+	}
+	return gc.client.SubmitValidatorRegistrations(gc.ctx, []*api.VersionedSignedValidatorRegistration{signedReg})
+}
+
+func (gc *goClient) SubmitValidatorRegistrationPostponed(pubkey []byte, feeRecipient bellatrix.ExecutionAddress, sig phase0.BLSSignature) error {
+	gc.postponedRegistrationsMu.Lock()
+	defer gc.postponedRegistrationsMu.Unlock()
+
+	currentSlot := uint64(gc.network.EstimatedCurrentSlot())
+	slotsPerEpoch := gc.network.SlotsPerEpoch()
+
+	shouldSubmit := len(gc.postponedRegistrations) != 0 &&
+		(currentSlot-gc.lastRegistrationSlot.Load() >= slotsPerEpoch &&
+			currentSlot%slotsPerEpoch == gc.operatorID%slotsPerEpoch ||
+			currentSlot-gc.lastRegistrationSlot.Load() >= slotsPerEpoch*2+gc.operatorID%slotsPerEpoch)
+
+	if shouldSubmit {
+		const batchSize = 500
+
+		for len(gc.postponedRegistrations) > 0 {
+			bs := batchSize
+			if bs > len(gc.postponedRegistrations) {
+				bs = len(gc.postponedRegistrations)
+			}
+
+			if err := gc.client.SubmitValidatorRegistrations(gc.ctx, gc.postponedRegistrations[0:bs]); err != nil {
+				return err
+			}
+
+			gc.log.Info("submitted postponed validator registrations", fields.Count(bs))
+
+			gc.postponedRegistrations = gc.postponedRegistrations[bs:]
+		}
+
+		gc.lastRegistrationSlot.Store(currentSlot)
+
+		return nil
+	}
+
+	pk := phase0.BLSPubKey{}
+	copy(pk[:], pubkey)
+	signedReg := &api.VersionedSignedValidatorRegistration{
+		Version: spec.BuilderVersionV1,
+		V1: &eth2apiv1.SignedValidatorRegistration{
+			Message: &eth2apiv1.ValidatorRegistration{
+				FeeRecipient: feeRecipient,
+				// TODO: This is a reasonable default, but we should probably make this configurable.
+				//       Discussion here: https://github.com/ethereum/builder-specs/issues/17
+				GasLimit:  30_000_000,
+				Timestamp: gc.network.GetSlotStartTime(gc.network.GetEpochFirstSlot(gc.network.EstimatedCurrentEpoch())),
+				Pubkey:    pk,
+			},
+			Signature: sig,
+		},
+	}
+
+	gc.postponedRegistrations = append(gc.postponedRegistrations, signedReg)
+
+	return nil
+}
+
+type ValidatorRegistration struct {
```

Comment:
seems to be unused.

Reply:
@y0sher Yes. This PR and the spec PR need a minor cleanup in the validator registration submission before being merged. `SubmitValidatorRegistrationBatched` is also to be removed.